streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -549,19 +559,101 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
+     * }</pre>
+     * <p>
+     * Within the {@link Transformer}, the state is obtained via the {@link ProcessorContext}.
+     * To trigger periodic actions via {@link org.apache.kafka.streams.processor.Punctuator#punctuate(long) punctuate()},
+     * a schedule must be registered.
+     *
+     * Method {@link Transformer#transform(Object, Object) transform()} must return an {@link java.util.Iterable} (e.g.,
+     * any {@link java.util.Collection} type) and the return value must not be {@code null}.
+     *
+     * <pre>{@code
+     * new TransformerSupplier() {
+     *     Transformer get() {
+     *         return new Transformer() {
+     *             private ProcessorContext context;
+     *             private StateStore state;
+     *
+     *             void init(ProcessorContext context) {
+     *                 this.context = context;
+     *                 this.state = context.getStateStore("myTransformState");
+     *                 // punctuate each 1000ms; can access this.state
+     *                 context.schedule(1000, PunctuationType.WALL_CLOCK_TIME, new Punctuator(..));
+     *             }
+     *
+     *             Iterable<KeyValue> transform(K key, V value) {
+     *                 // can access this.state
+     *                 List<KeyValue> result = new ArrayList<>();
+     *                 for (int i = 0; i < n; i++) {
+     *                     result.add(new KeyValue(key, value));
+     *                 }
+     *                 return result; // emits a list of key-value pairs via return
+     *             }
+     *
+     *             void close() {
+     *                 // can access this.state
+     *             }
+     *         }
+     *     }
+     * }
+     * }</pre>
+     * <p>
+     * Transforming records might result in an internal data redistribution if a key based operator (like an aggregation
+     * or join) is applied to the result {@code KStream}.
+     * (cf. {@link #transformValues(ValueTransformerSupplier, String...)})
+     *
+     * @param transformerSupplier an instance of {@link TransformerSupplier} that generates a {@link Transformer}
+     * @param stateStoreNames     the names of the state stores used by the processor
+     * @param <K1>                the key type of the new stream
+     * @param <V1>                the value type of the new stream
+     * @return a {@code KStream} that contains more or less records with new key and value (possibly of different type)
+     * @see #flatMap(KeyValueMapper)
+     * @see #transform(TransformerSupplier, String...)
+     * @see #transformValues(ValueTransformerSupplier, String...)
+     * @see #transformValues(ValueTransformerWithKeySupplier, String...)
+     * @see #process(ProcessorSupplier, String...)
+     */
+    <K1, V1> KStream<K1, V1> flatTransform(final TransformerSupplier<? super K, ? super V, ? extends Iterable<? extends KeyValue<? extends K1, ? extends V1>>> transformerSupplier,
```

Comment:
Generics confuse me regularly, too... I was just comparing to `transform` -- seems our API is not consistent. I guess your argument makes sense.

Reply:
Regarding `<K1, V1>` vs. `<? extends K1, ? extends V1>`, I changed my mind. I guess also here `<? extends K1, ? extends V1>` makes sense because a user could also give K1 and V1 explicitly.

Reply:
I hope you don't mind if I jump in...

It's probably worth doing some experiments to get the bounds just right. Here are the basic properties we need:

Let's say we have a simple class hierarchy:
```
interface Animal
interface Cat extends Animal
interface Dog extends Animal
```
And let's say we subclass `KeyValue`: `public class KV extends KeyValue`.

Given a `KStream<String, Cat>`, we want to be able to call `flatTransform` with a `TransformerSupplier<String, Animal, List<KeyValue<Animal>>>` like this one:
`(String s, Animal a) -> asList(new KV(s, new Cat()), new KeyValue(s, new Dog()))`.

The `? super K` and `? super V` ensure that any transformer that _can handle_ a `K` and `V` is permitted. You _can handle_ an instance if you can assign it to you parameter types (so your parameter type is a supertype of the instance type). You need to specify this as a bound because the compiler doesn't generally let me assign a supertype to a subtype (I can't write `Cat c = (Animal) a;`).

I don't think you actually need the `? extends Whatever` parts to get the right outcome. If anything, you might need it on the stuff inside the iterable to make sure that you can just pass in heterogeneous members, and they'll get "rounded" to their lowest common superclass.

I think this works without `extends` because you generally _can_ assign a subtype to a supertype.

So maybe try an example like this without the `extends` stuff and see if the compiler chokes on it or not.

Reply:
IIRC, Java doesn't fully implement variance, but this is the basic concept (in scala) of what's going on with those `extends` return types: https://docs.scala-lang.org/tour/variances.html

Reply:
Ok, I stand corrected. I did the experiment I was suggesting, and I got the desired results only with the API you have submitted:

```java
<K1, V1> KStreamImpl<K1, V1> flatTransform(final TransformerSupplier<? super K, ? super V, ? extends Iterable<? extends KeyValue<? extends K1, ? extends V1>>> transformerSupplier,
                                               String... storeNames) {
        return null;
    }

    public static class KV<K, V> extends KeyValue<K, V> {
        public KV(final K key, final V value) {
            super(key, value);
        }
    }

    public static void main(String[] args) {

        final KStreamImpl<Integer, Long> stream = new KStreamImpl<>(null, null, null, false, null);

        // exact transformer
        final KStreamImpl<Integer, Long> stream2 = stream.flatTransform(
            new TransformerSupplier<Integer, Long, Iterable<KeyValue<Integer, Long>>>() {
                @Override
                public Transformer<Integer, Long, Iterable<KeyValue<Integer, Long>>> get() {
                    return new Transformer<Integer, Long, Iterable<KeyValue<Integer, Long>>>() {
                        @Override
                        public void init(final ProcessorContext context) {}

                        @Override
                        public Iterable<KeyValue<Integer, Long>> transform(final Integer key, final Long value) {
                            return Arrays.asList(new KV<>(key, value), new KeyValue<>(key, value));
                        }

                        @Override
                        public void close() {}
                    };
                }
            }
        );

        // transformer that takes superclass k/v and returns exact results
        final KStreamImpl<Integer, Long> stream3 = stream.flatTransform(
            new TransformerSupplier<Number, Number, Iterable<KeyValue<Integer, Long>>>() {
                @Override
                public Transformer<Number, Number, Iterable<KeyValue<Integer, Long>>> get() {
                    return new Transformer<Number, Number, Iterable<KeyValue<Integer, Long>>>() {
                        @Override
                        public void init(final ProcessorContext context) {

                        }

                        @Override
                        public Iterable<KeyValue<Integer, Long>> transform(final Number key, final Number value) {
                            return Arrays.asList(new KV<>(key.intValue(), value.longValue()), new KeyValue<>(1, 3L));
                        }

                        @Override
                        public void close() {

                        }
                    };
                }
            }
        );


        // transformer that takes exact parameters and returns subclass results
        final KStreamImpl<Number, Number> stream4 = stream.flatTransform(
            new TransformerSupplier<Integer, Long, Iterable<KeyValue<Integer, Long>>>() {
                @Override
                public Transformer<Integer, Long, Iterable<KeyValue<Integer, Long>>> get() {
                    return new Transformer<Integer, Long, Iterable<KeyValue<Integer, Long>>>() {
                        @Override
                        public void init(final ProcessorContext context) {

                        }

                        @Override
                        public Iterable<KeyValue<Integer, Long>> transform(final Integer key, final Long value) {
                            return Arrays.asList(new KV<>(key, value), new KeyValue<>(1, 3L));
                        }

                        @Override
                        public void close() {

                        }
                    };
                }
            }
        );


        // transformer that takes superclass parameters and returns subclass results
        final KStreamImpl<Number, Number> stream5 = stream.flatTransform(
            new TransformerSupplier<Number, Number, Iterable<KeyValue<Integer, Long>>>() {
                @Override
                public Transformer<Number, Number, Iterable<KeyValue<Integer, Long>>> get() {
                    return new Transformer<Number, Number, Iterable<KeyValue<Integer, Long>>>() {
                        @Override
                        public void init(final ProcessorContext context) {

                        }

                        @Override
                        public Iterable<KeyValue<Integer, Long>> transform(final Number key, final Number value) {
                            return Arrays.asList(new KV<>(key.intValue(), value.longValue()), new KeyValue<>(1, 3L));
                        }

                        @Override
                        public void close() {

                        }
                    };
                }
            }
        );

    }
```

If you take out any of the `? extends`, it won't compile.

Reply:
Thanks for verifying @vvcephei!